examples/gdp/eight_process/eight_proc_logical.py

```diff
@@ -0,0 +1,202 @@
+"""Disjunctive re-implementation of eight-process problem.
+
+Re-implementation of Duran example 3 superstructure synthesis problem in Pyomo
+with full leverage of GDP framework.
+
+This is a convex MINLP problem that has been formulated as a GDP model. The
+expected optimal solution value is 68.0.
+
+Ref:
+    SELECT OPTIMAL PROCESS FROM WITHIN GIVEN SUPERSTRUCTURE.
+    MARCO DURAN , PH.D. THESIS (EX3) , 1984.
+    CARNEGIE-MELLON UNIVERSITY , PITTSBURGH , PA.
+
+(original problem, my implementation may vary)
+    Problem type:    convex MINLP
+            size:    8  binary variables
+                    26  continuous variables
+                    32  constraints
+
+Pictoral representation can be found on
+page 969 of Turkay & Grossmann, 1996.
+http://dx.doi.org/10.1016/0098-1354(95)00219-7
+
+"""
+from __future__ import division
+
+from six import iteritems
+
+from pyomo.core.expr.logical_expr import land
+from pyomo.environ import (ConcreteModel, Constraint, NonNegativeReals,
+                           Objective, Param, RangeSet, Var, exp, minimize, BooleanVar, LogicalConstraint)
+from pyomo.gdp import Disjunction
+from pyomo.opt import SolverFactory
+
+
+def build_eight_process_flowsheet():
+    """Build flowsheet for the 8 process problem."""
+    m = ConcreteModel(name='DuranEx3 Disjunctive')
+
+    """Set declarations"""
+    m.streams = RangeSet(2, 25, doc="process streams")
+    m.units = RangeSet(1, 8, doc="process units")
+
+    """Parameter and initial point declarations"""
+    # FIXED COST INVESTMENT COEFF FOR PROCESS UNITS
+    # Format: process #: cost
+    fixed_cost = {1: 5, 2: 8, 3: 6, 4: 10, 5: 6, 6: 7, 7: 4, 8: 5}
+    m.CF = Param(m.units, initialize=fixed_cost)
+
+    def fixed_cost_bounds(m, unit):
+        return (0, m.CF[unit])
+    m.yCF = Var(m.units, initialize=0, bounds=fixed_cost_bounds)
+
+    # VARIABLE COST COEFF FOR PROCESS UNITS - STREAMS
+    # Format: stream #: cost
+    variable_cost = {3: -10, 5: -15, 9: -40, 19: 25, 21: 35, 25: -35,
+                     17: 80, 14: 15, 10: 15, 2: 1, 4: 1, 18: -65, 20: -60,
+                     22: -80}
+    CV = m.CV = Param(m.streams, initialize=variable_cost, default=0)
+
+    # initial point information for stream flows
+    initX = {2: 2, 3: 1.5, 6: 0.75, 7: 0.5, 8: 0.5, 9: 0.75, 11: 1.5,
+             12: 1.34, 13: 2, 14: 2.5, 17: 2, 18: 0.75, 19: 2, 20: 1.5,
+             23: 1.7, 24: 1.5, 25: 0.5}
+
+    """Variable declarations"""
+    # FLOWRATES OF PROCESS STREAMS
+    m.flow = Var(m.streams, domain=NonNegativeReals, initialize=initX,
+                 bounds=(0, 10))
+    m.use_unit = BooleanVar(m.units)
+    # OBJECTIVE FUNCTION CONSTANT TERM
+    CONSTANT = m.constant = Param(initialize=122.0)
+
+    """Constraint definitions"""
+    # INPUT-OUTPUT RELATIONS FOR process units 1 through 8
+    m.use_unit_1or2 = Disjunction(
+        expr=[
+            # use unit 1 disjunct
+            [m.use_unit[1], ~m.use_unit[2],
```

Comment:
Yeah, for implicitly defined Disjuncts, I wasn't sure exactly how I wanted to "hack" the fact that Disjunct indicators are binary rather than Boolean.